### DISCUSSION TITLE
How run next.js with node-sharp for docker

### DISCUSSION THREAD

**Discussion by miko866** [vercel/next.js#35296] [Help] [answered]
**I have a problem implementing sharp for Dockerfile.** 

**Error: 'sharp' is required to be installed in standalone mode for the image optimization to function correctly**

Next.js with sharp works fine for local developing:
 - next 12.0.1
 - sharp 0.30.2
 - node 16.xx
 - npm 8.xx
 - OS - macOS Monterey - 12.2.1, M1 PRO

**next.config.js**

    module.exports = {
      experimental: {
        outputStandalone: true,
      },

}

**Dockerfile:**

    FROM node:16-alpine AS deps
    # Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
    RUN apk add --no-cache libc6-compat
    
    WORKDIR /app
    COPY package.json package-lock.json ./ 
    RUN npm ci
    
    # Rebuild the source code only when needed
    FROM node:16-alpine AS builder
    WORKDIR /app
    COPY --from=deps /app/node_modules ./node_modules
    COPY . .
    
    # Next.js collects completely anonymous telemetry data about general usage.
    # Learn more here: https://nextjs.org/telemetry
    # Uncomment the following line in case you want to disable telemetry during the build.
    # ENV NEXT_TELEMETRY_DISABLED 1
    
    RUN npm run build
    
    # Production image, copy all the files and run next
    FROM node:16-alpine AS runner
    WORKDIR /app
    
    ENV NODE_ENV production
    # Uncomment the following line in case you want to disable telemetry during runtime.
    # ENV NEXT_TELEMETRY_DISABLED 1
    
    RUN addgroup --system --gid 1001 nodejs
    RUN adduser --system --uid 1001 nextjs
    
    # You only need to copy next.config.js if you are NOT using the default configuration
    COPY --from=builder /app/public ./public
    COPY --from=builder /app/package.json ./package.json
    COPY --from=builder /app/package.json ./package.json
    COPY --from=builder /app/next.config.js ./
    COPY --from=builder /app/next-i18next.config.js ./
    COPY --from=builder /app/next-sitemap.js ./cd
    COPY --from=builder /app/jsconfig.json ./jsconfig.json
    COPY --from=builder /app/data/ ./data
    COPY --from=builder /app/components/ ./components
    COPY --from=builder /app/utils/ ./utils
    COPY --from=builder /app/assets/ ./assets
    
    # Automatically leverage output traces to reduce image size 
    # https://nextjs.org/docs/advanced-features/output-file-tracing
    COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
    COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static
    
    USER nextjs
    
    EXPOSE 3000
    
    ENV PORT 3000
    
    CMD ["node", "server.js"]

**.env file:**

    NEXT_SHARP_PATH=/tmp/node_modules/sharp next start

Sharp is installed in package.json

I checked both Next/Vercel tuts:
- [https://nextjs.org/docs/messages/install-sharp][1]
- [https://nextjs.org/docs/messages/sharp-missing-in-production][2]

**RUN Docker:** 

    docker build --no-cache . -t website-app && docker run --name website -p 3000:3000 website-app

  [1]: https://nextjs.org/docs/messages/install-sharp
  [2]: https://nextjs.org/docs/messages/sharp-missing-in-production

**Comment by wwwdepot** [comment, marked as answer]
> Sharp is installed in package.json

If you have installed sharp in package.json, you do not need to specify sharp location in the `.env` file.

**Reply by miko866** [reply]
I tried it before and it doesn't work. 
Docker is build and started but as soon as I open the address in Browser, the docker app stops without error.

**Reply by wwwdepot** [reply]
My setup is using [with-docker](https://github.com/vercel/next.js/tree/canary/examples/with-docker) example.

My Steps:
1. Install sharp `npm i sharp`
2. create a new page with Image component 
```jsx
import Image from "next/image";

export default function IndexPage() {
  return (
    <div>
      <h1>image</h1>
      <Image
        width={750}
        height={449}
        src={
          "https://images.unsplash.com/photo-1556911220-bff31c812dba?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop"
        }
        layout="responsive"
      />
    </div>
  );
}
```
3. configure `next.config.js` to use with Image component
```js
// next.config.js
module.exports = {
  experimental: {
    outputStandalone: true,
  },
  images: {
    imageSizes: [16, 32, 48, 64, 96, 128, 256],
    domains: ["images.unsplash.com"],
    minimumCacheTTL: 3600,
    formats: ["image/webp"],
  },
}
```
4. build container
5. run container. No error.

Your `next.config.js` file is not configured with images. Please note **When using an external URL, you must add it to domains in next.config.js.**

**Reply by miko866** [reply]
I host all images in - public/images.

**Comment by miko866** [comment]
I figured it out.

I removed `NEXT_SHARP_PATH=/tmp/node_modules/sharp next start` from .env
Local docker doesn't work with node-sharp without `NEXT_SHARP_PATH` and that is weird. 

But I deployed it in my K8s Cluster with Docker and it works as is expected.

**Reply by smo043** [reply]
Hello, I tried this solution still see an error when deployed docker image to Azure app service.Its failing with or without NEXT_SHARP_PATH. Can someone pls help.

Error: 'sharp' is required to be installed in standalone mode for the image optimization to function correctly.

**Reply by Smarello** [reply]
Same issue here.
I'm deploying on an internal kubernetes cluster and no matter if I set the `NEXT_SHARP_PATH` variable or not, I always get the error

> Error: 'sharp' is required to be installed in standalone mode for the image optimization to function correctly

I'm using these versions in the `package.json`
"next": "12.0.9",
"sharp": "0.30.4"

**Reply by smo043** [reply]
Team - any help here?

**Reply by RobertJif** [reply]
Found any solution ?
having the same problem, deploying to google kubernetes cluster.

**Comment by CodeblackNL** [comment]
Anyone find a solution yet?

Same issue here.
`Error: 'sharp' is required to be installed in standalone mode for the image optimization to function correctly`

Using:
next: 12.2.5
sharp: 0.30.7
npm: 8.15.0

`sharp` is installed. I've checked in the container, it's there in node_modules.
Setting `NEXT_SHARP_PATH` makes no difference.
I've also tried other `node` base-images i.s.o. `node-alpine`, like the debian or buster-slim variants, as I read that alpine & sharp don't work together, but makes no difference.
And tried al kinds of changes to `next.config.js`, that I found here and there, too many to mention, but also makes no difference.

All I find is a lot of people struggling with this, but never a solution.
Any help or suggestions are highly appreciated.

**Reply by CodeblackNL** [reply]
As it turns out, I found the solution myself :-)
It is the `--ignore-scripts` that is part of the `npm install` command.

Change this:
`RUN npm install --production --frozen-lockfile --ignore-scripts --prefer-offline`
to this:
`RUN npm install --production --frozen-lockfile --prefer-offline`

**Reply by eudueu** [reply]
What works for me, is to change the version of Alpine Linux in the Dockerfile:
`FROM node:18-alpine AS deps` (instead of node:16)

I didn't add node-sharp to my `.env` file.

**Reply by giovanifranz** [reply]
solved my problem

**Comment by JunlinPan-Crypto** [comment]
I figured it out!

The reason is that I used different architecture and platform of Node.js at build and runtime. My dockerfile:
```
# alpine version here
FROM node:18.10.0-alpine AS build
RUN npm build

# base linux version
FROM nginx:1.23.1 AS runtime
RUN curl -sL https://deb.nodesource.com/setup_18.x | bash
RUN apt-get update && apt-get install -y nodejs
...
```
Ref: https://github.com/lovell/sharp/issues/2312

> The architecture and platform of Node.js used for npm install must be the same as the architecture and platform of Node.js used at runtime.

The sharp bin version built in first build stage doesn't match `sharp-linux-x64` version required in runtime:

```
# ls build/Release
sharp-linuxmusl-x64.node
```

<img width="1105" alt="image" src="https://user-images.githubusercontent.com/90383710/195047642-cb20e108-a790-4c91-9244-57777f55c227.png">


So I just use the `slim` node instead of `alpine` version.
```
FROM node:18.10.0-slim AS build
```

Hope it helps 😄

**Comment by spiderneo** [comment]
I had the same issue.

I was doing my dev on macOS, dev and build was OK on Mac. I made a tar of the docker image on my Mac, and load it on a prod server. Impossible to have image optimization.
But if i built it on the server directly, no issue, all is working well.

**Comment by coderNeos** [comment]
The problem is that the example in [with-docker](https://github.com/vercel/next.js/tree/canary/examples/with-docker) repo is incorrect. 
```Dockerfile
COPY --from=deps /app/node_modules ./node_modules
COPY . .
```
First it copies node_modules from deps and then it overrides them by copying from local directory `COPY . .`

There is my version of slightly modified Dockerfile which works correctly:
```Dockerfile
# Install dependencies only when needed
FROM node:16-alpine AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i --frozen-lockfile; \
  else echo "Lockfile not found." && exit 1; \
  fi


# Rebuild the source code only when needed
FROM node:16-alpine AS builder
WORKDIR /app
COPY pages/ ./pages/
COPY public/ ./public/
COPY server/ ./server/
COPY src/ ./src/
COPY .eslintignore            \
     .eslintrc.json           \
     .gitignore               \
     next-env.d.ts            \
     next.config.js           \
     tsconfig.json            \
     ./
COPY --from=deps /app/package.json ./package.json
COPY --from=deps /app/package-lock.json ./package-lock.json
COPY --from=deps /app/node_modules ./node_modules
# Add env for production
# COPY .docker/production/.env.local .env.local

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry during the build.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN npm run build

# Production image, copy all the files and run next
FROM node:16-alpine AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/.env.local ./.env.local
COPY --from=builder /app/public ./public

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000

CMD ["node", "server.js"]
```

P.s. also note that it copies only the required dirs and files from local directory to speed up the build

**Reply by sz7i** [reply]
Thanks for this! I think the root cause of this issue is correctly identified and it's related to overwriting ./node_modules from local. In my case I solved it by adding the missing .dockerignore from the mentioned above repository [with-docker](https://github.com/vercel/next.js/tree/canary/examples/with-docker)

**Reply by paales** [reply]
To make it simpler, the two lines need to be flipped:

```
COPY . .
COPY --from=deps /app/node_modules ./node_modules
```

Intead of
https://github.com/vercel/next.js/blob/8bacad921be0f8cbf88a1e1d4ee12fbdf54faa6e/examples/with-docker/Dockerfile#L22-L23

**Reply by singleseeker** [reply]
```sh
Failed to write image to cache 36jF-z9kKkqsbyDI8ij7RSRmm9C5uJ2c1ubArSfeew8= [Error: EACCES: permission denied, mkdir '/app/.next/cache'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'mkdir',
  path: '/app/.next/cache'
}
```

Even if I flipped these two lines, I still get the errors.

**Reply by zhdanovartur** [reply]
One more solution is to install `sharp` with specific args:

```dockerfile
RUN npm install -g --arch=x64 --platform=linux --libc=glibc sharp
```

and then copy this global package into `runner`:

```dockerfile
ENV NEXT_SHARP_PATH=/usr/local/lib/node_modules/sharp
COPY --from=deps --chown=nextjs:nodejs /usr/local/lib/node_modules/sharp /usr/local/lib/node_modules/sharp
```

Example:

```dockerfile
FROM node:21.4-alpine AS deps
RUN apk add --no-cache libc6-compat
WORKDIR /app

COPY package.json ./
COPY package-lock.json ./

RUN npm ci

RUN npm install -g --arch=x64 --platform=linux --libc=glibc sharp

FROM node:21.4-alpine AS builder
WORKDIR /app

COPY . .
COPY --from=deps /app/node_modules ./node_modules

RUN npm run build

FROM node:21.4-alpine AS runner
WORKDIR /app

ENV NODE_ENV production
ENV NEXT_TELEMETRY_DISABLED 1
ENV NEXT_SHARP_PATH=/usr/local/lib/node_modules/sharp

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=deps --chown=nextjs:nodejs /usr/local/lib/node_modules/sharp /usr/local/lib/node_modules/sharp
COPY --from=builder /app/public ./public
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

COPY .env.production /app/.env.production
RUN chown -R nextjs:nodejs /app/.env.production

USER nextjs

EXPOSE 8080

ENV PORT 8080

CMD ["node", "server.js"]

```

**Reply by darincon** [reply]
The answer from @zhdanovartur worked like a charm. I had a hard time making the solution from @coderNeos work. Quick question, is it ok to do `npm install -g` inside docker? I know it's not recommended on your local but I'm wondering whether it is a good practice or not. Any thoughts on this? I'm new with docker and VMs in general

**Reply by zhdanovartur** [reply]
@darincon Good point. Overall, it's nothing serious I believe, but indeed a local installation is always preferable to a global one. Unfortunately, I don't remember if I did this deliberately or if there were any issues. Let us know please if local one works well.

**Reply by bharad22** [reply]
This code is working fine, but im using yarn
do you know what changes are required? and what will be the path of node_modules?

**Reply by bharad22** [reply]
I just found that we don't need to specify NEXT_SHARP_PATH in env and don't need to install sharp by specifying --arch, --platform, --libc. We can directly install sharp by specifying --ignore-engines.

```
FROM node:20-alpine

WORKDIR /out
ARG configuration

COPY package*.json yarn.lock ./

RUN yarn install
RUN yarn add sharp --ignore-engines

COPY . .
RUN cp /out/${configuration} .env

RUN yarn build

EXPOSE 3000

**Reply by deadcoder0904** [reply]
@bharad22 thanks for your solution. [i implemented your solution in `pnpm`](https://github.com/vercel/next.js/discussions/35296#discussioncomment-8626288)

**Comment by viantirreau** [comment]
I had the same issue and couldn't figure out how to solve it with `pnpm` as a package manager. For reference, I'm using `next@14.1.0`.

Turns out that the equivalent to the command
```dockerfile
RUN npm install -g --arch=x64 --platform=linux --libc=glibc sharp
```

with `pnpm` is 
```dockerfile
RUN pnpm i -g --config.arch=x64 --config.platform=linux --config.libc=glibc sharp
```

After more tinkering, pinning the `sharp` version to an older one and just installing it normally instead of globally, I thought about running `node` directly inside the docker container and try to `require` sharp (as in the [source code](https://github.com/vercel/next.js/blob/da616e56d265d2db1c9808129bd28c6d1884d3a9/packages/next/src/server/image-optimizer.ts#L50)).
I got the following (very informative) message:

```bash
/app $ node
Welcome to Node.js v18.19.1.
Type ".help" for more information.
> sharp = require("/app/node_modules/sharp")
Uncaught Error: 
Something went wrong installing the "sharp" module

Cannot find module '../build/Release/sharp-linuxmusl-x64.node'
Require stack:
- /app/node_modules/.pnpm/sharp@0.32.6/node_modules/sharp/lib/sharp.js
- /app/node_modules/.pnpm/sharp@0.32.6/node_modules/sharp/lib/constructor.js
- /app/node_modules/.pnpm/sharp@0.32.6/node_modules/sharp/lib/index.js
- <repl>

Possible solutions:
- Install with verbose logging and look for errors: "npm install --ignore-scripts=false --foreground-scripts --verbose sharp"
- Install for the current linuxmusl-x64 runtime: "npm install --platform=linuxmusl --arch=x64 sharp"
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
```
Turns out that I'm using alpine linux, so apparently forcing the usage of `glibc` breaks the installation, because the runner expects a `musl` version. Additionally, as suggested in this discussion https://github.com/vercel/next.js/discussions/40125#discussioncomment-8254222, I opted to install a pinned version instead of the latest one. Thus, my `RUN` command was the following:

```dockerfile
RUN pnpm i --config.arch=x64 --config.platform=linux --config.libc=musl sharp@0.32.6
```

Notice the removal of the global flag, the specific `libc` and the version pinning.

After confirming the fix, I tried to go back to the latest version (0.33.2) by removing the version pinning, but it broke again with that single change, confirming what was suggested by @vvbbnn00, who links back to #59516.

For reference, my final Dockerfile looks like the following:

<details>
  <summary>Expand Dockerfile</summary>
  
 
  ```dockerfile
# https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile
FROM node:18-alpine AS base

# Install dependencies only when needed
FROM base AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
# ENV PNPM_HOME=/usr/local/bin
RUN corepack enable pnpm && pnpm i --frozen-lockfile
RUN pnpm i --config.arch=x64 --config.platform=linux --config.libc=musl sharp@0.32.6

# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
# https://github.com/vercel/next.js/discussions/36935
# Uncomment the following line if there are permission issues with .next/cache directory
# RUN mkdir -p /app/.next/cache && chown nextjs:nodejs /app/.next/cache
RUN mkdir -p /app/.next/cache
# Persist the next cache in a volume
VOLUME ["/app/.next/cache"]
COPY . .
COPY --from=deps /app/node_modules ./node_modules

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry during the build.
ENV NEXT_TELEMETRY_DISABLED 1

ENV NEXT_SHARP_PATH=/app/node_modules/sharp

RUN corepack enable pnpm && pnpm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

# COPY --from=deps --chown=nextjs:nodejs /usr/local/lib/node_modules/sharp /usr/local/lib/node_modules/sharp
COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000
# set hostname to localhost
ENV HOSTNAME "0.0.0.0"
# server.js is created by next build from the standalone output
# https://nextjs.org/docs/pages/api-reference/next-config-js/output
CMD ["node", "server.js"]
  ```
  
</details>

**Reply by deadcoder0904** [reply]
thanks @viantirreau, i faced the same error 2 weeks later. also, found out to save next cache so thanks for sharing the dockerfile as well.

Bdw, I just added `NEXT_SHARP_PATH` environment variable before `pnpm build` (`next build`) happens. I've already installed `sharp` as a dependency in my `package.json` bdw.

Turns out, commenting `RUN pnpm install --config.arch=x64 --config.platform=linux --config.libc=musl sharp@0.33.2` doesn't work.

### Dockerfile

```bash
RUN pnpm install --config.arch=x64 --config.platform=linux --config.libc=musl sharp@0.33.2

ENV NEXT_SHARP_PATH=/app/node_modules/sharp
RUN pnpm build
```

Next.js doc has [a simple guide on it here](https://nextjs.org/docs/messages/sharp-missing-in-production).

**Reply by viantirreau** [reply]
Glad that it worked!

> Turns out, commenting RUN pnpm install --config.arch=x64 --config.platform=linux --config.libc=musl sharp@0.33.2 doesn't work.

I ended up removing `sharp` as a dependency in my `package.json`, because my system's `libc` is `glibc`, while `alpine` uses `musl`, so I was kind of forced to side load it just on the Dockerfile. 
I believe that's why commenting that line doesn't work: it leaves you with the `sharp` version indicated by the `package.json`, which is incompatible with `alpine`. Because it has to be overridden anyways, I opted to remove `sharp` from the `package.json` altogether.

**Reply by deadcoder0904** [reply]
@viantirreau it didn't work. i spoke way too soon. i'm still trying to make it work. i removed `sharp` too.

my current error:

```bash
/app $ node
Welcome to Node.js v20.11.1.
Type ".help" for more information.
> sharp = require("/app/node_modules/sharp")
Uncaught Error: Could not load the "sharp" module using the linuxmusl-x64 runtime
ERR_DLOPEN_FAILED: Error loading shared library libvips-cpp.so.42: No such file or directory (needed by /app/node_modules/.pnpm/@img+sharp-linuxmusl-x64@0.33.2/node_modules/@img/sharp-linuxmusl-x64/lib/sharp-linuxmusl-x64.node)
Possible solutions:
- Ensure optional dependencies can be installed:
    npm install --include=optional sharp
    yarn add sharp --ignore-engines
- Ensure your package manager supports multi-platform installation:
    See https://sharp.pixelplumbing.com/install#cross-platform
- Add platform-specific dependencies:
    npm install --os=linux --libc=musl --cpu=x64 sharp
- Consult the installation documentation:
    See https://sharp.pixelplumbing.com/install
    at Object.<anonymous> (/app/node_modules/.pnpm/sharp@0.33.2/node_modules/sharp/lib/sharp.js:114:9)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
```

i came to same conclusion as you for removing the sharp part.

can you find faults in my dockerfile?

#### Dockerfile

```dockerfile
FROM node:20-alpine AS base

# mostly inspired from https://github.com/BretFisher/node-docker-good-defaults/blob/main/Dockerfile & https://github.com/remix-run/example-trellix/blob/main/Dockerfile

# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
RUN corepack enable && corepack prepare pnpm@8.15.4 --activate 
# set the store dir to a folder that is not in the project
RUN pnpm config set store-dir ~/.pnpm-store
RUN pnpm fetch

# 1. Install all dependencies including dev dependencies
FROM base AS deps
# Root user is implicit so you don't have to actually specify it. From https://stackoverflow.com/a/45553149/6141587
# USER root
USER node
# WORKDIR now sets correct permissions if you set USER first so `USER node` has permissions on `/app` directory
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY --chown=node:node package.json pnpm-lock.yaml* ./
COPY --chown=node:node /src/app/db/migrations ./migrations

USER root
RUN pnpm install --frozen-lockfile --prefer-offline
RUN pnpm install sharp

# 2. Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
# Inspired by https://github.com/vercel/next.js/discussions/36935
RUN mkdir -p /app/.next/cache && chown -R node:node /app/.next/cache
# Persist the next cache in a volume
VOLUME ["/app/.next/cache"]
COPY --from=deps --chown=node:node /app/node_modules ./node_modules

COPY --chown=node:node . .

# This will do the trick, use the corresponding env file for each environment.
COPY --chown=node:node .env.production .env.production

# Copied from https://stackoverflow.com/a/69867550/6141587
USER root
# Give /data directory correct permissions otherwise WAL mode won't work. It means you can't have 2 users writing to the database at the same time without this line as *.sqlite-wal & *.sqlite-shm are automatically created & deleted when *.sqlite is busy.
RUN mkdir -p /data && chown -R node:node /data

ENV NODE_ENV=production
ENV NEXT_TELEMETRY_DISABLED=1
# ENV NEXT_SHARP_PATH=/app/node_modules/sharp

RUN pnpm build

# 3. Production image, copy all the files and run next
FROM base AS runner
USER node
WORKDIR /app

EXPOSE 3000

ENV PORT 3000
ENV HOSTNAME 0.0.0.0
ENV NODE_ENV=production
ENV NEXT_TELEMETRY_DISABLED=1
# ENV NEXT_SHARP_PATH=/app/node_modules/sharp

COPY --from=builder --chown=node:node /app/public ./public

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=node:node /app/.next/standalone ./
COPY --from=builder --chown=node:node /app/.next/static ./.next/static

# Move the drizzle directory to the runtime image
COPY --from=builder --chown=node:node /app/src/app/db/migrations ./migrations

# Move the litestream binary to the runtime image from the litestream image
# You can use a specific version of litestream by changing the tag
# COPY --from=litestream/litestream:0.3.13 /usr/local/bin/litestream /usr/local/bin/litestream
COPY --from=litestream/litestream:latest --chown=node:node /usr/local/bin/litestream /usr/local/bin/litestream
COPY --from=builder --chown=node:node /app/litestream.yml /etc/litestream.yml

COPY --from=builder --chown=node:node /app/scripts/drizzle-migrate.mjs ./scripts/drizzle-migrate.mjs
COPY --from=builder --chown=node:node /app/scripts/package.json ./scripts/package.json
COPY --from=builder --chown=node:node /app/scripts/pnpm-lock.yaml ./scripts/pnpm-lock.yaml
COPY --from=builder --chown=node:node /app/scripts/run.sh ./run.sh
RUN chmod +x run.sh

CMD ["sh", "run.sh"]
```

i even tried adding this in `package.json` because [sharp recommends it here](https://sharp.pixelplumbing.com/install#cross-platform):

```package.json
"pnpm": {
		"supportedArchitectures": {
			"os": [
				"current"
			],
			"cpu": [
				"x64",
				"arm64"
			]
		}
	}
```

**Reply by deadcoder0904** [reply]
alright, i finally solved it.

i didn't do anything different other than deleting `node_modules` & reinstalling dependencies again.

the solution with `pnpm` was to normally install it afterwards:

```dockerfile
RUN pnpm install --frozen-lockfile --prefer-offline
RUN pnpm install sharp@0.32.6
```

i think putting this in my `package.json` helped:

```json
	"pnpm": {
		"supportedArchitectures": {
			"os": [
				"current"
			],
			"cpu": [
				"x64",
				"arm64"
			]
		}
	}
```

my dockerfile doesn't even contain environment variable `NEXT_SHARP_PATH` now & it works:

```dockerfile
FROM node:20-alpine AS base

# mostly inspired from https://github.com/BretFisher/node-docker-good-defaults/blob/main/Dockerfile & https://github.com/remix-run/example-trellix/blob/main/Dockerfile

# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
RUN corepack enable && corepack prepare pnpm@8.15.4 --activate 
# set the store dir to a folder that is not in the project
RUN pnpm config set store-dir ~/.pnpm-store
RUN pnpm fetch

# 1. Install all dependencies including dev dependencies
FROM base AS deps
# Root user is implicit so you don't have to actually specify it. From https://stackoverflow.com/a/45553149/6141587
# USER root
USER node
# WORKDIR now sets correct permissions if you set USER first so `USER node` has permissions on `/app` directory
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY --chown=node:node package.json pnpm-lock.yaml* ./
COPY --chown=node:node /src/app/db/migrations ./migrations

USER root
RUN pnpm install --frozen-lockfile --prefer-offline
RUN pnpm install sharp@0.32.6

# 2. Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
# Inspired by https://github.com/vercel/next.js/discussions/36935
RUN mkdir -p /app/.next/cache && chown -R node:node /app/.next/cache
# Persist the next cache in a volume
VOLUME ["/app/.next/cache"]
COPY --from=deps --chown=node:node /app/node_modules ./node_modules

COPY --chown=node:node . .

# This will do the trick, use the corresponding env file for each environment.
COPY --chown=node:node .env.production .env.production

# Copied from https://stackoverflow.com/a/69867550/6141587
USER root
# Give /data directory correct permissions otherwise WAL mode won't work. It means you can't have 2 users writing to the database at the same time without this line as *.sqlite-wal & *.sqlite-shm are automatically created & deleted when *.sqlite is busy.
RUN mkdir -p /data && chown -R node:node /data

ENV NODE_ENV=production
ENV NEXT_TELEMETRY_DISABLED=1

RUN pnpm build

# 3. Production image, copy all the files and run next
FROM base AS runner
USER node
WORKDIR /app

EXPOSE 3000

ENV PORT 3000
ENV HOSTNAME 0.0.0.0
ENV NODE_ENV=production
ENV NEXT_TELEMETRY_DISABLED=1

COPY --from=builder --chown=node:node /app/public ./public

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=node:node /app/.next/standalone ./
COPY --from=builder --chown=node:node /app/.next/static ./.next/static

# Move the drizzle directory to the runtime image
COPY --from=builder --chown=node:node /app/src/app/db/migrations ./migrations

# Move the litestream binary to the runtime image from the litestream image
# You can use a specific version of litestream by changing the tag
# COPY --from=litestream/litestream:0.3.13 /usr/local/bin/litestream /usr/local/bin/litestream
COPY --from=litestream/litestream:latest --chown=node:node /usr/local/bin/litestream /usr/local/bin/litestream
COPY --from=builder --chown=node:node /app/litestream.yml /etc/litestream.yml

COPY --from=builder --chown=node:node /app/scripts/drizzle-migrate.mjs ./scripts/drizzle-migrate.mjs
COPY --from=builder --chown=node:node /app/scripts/package.json ./scripts/package.json
COPY --from=builder --chown=node:node /app/scripts/pnpm-lock.yaml ./scripts/pnpm-lock.yaml
COPY --from=builder --chown=node:node /app/scripts/run.sh ./run.sh
RUN chmod +x run.sh

CMD ["sh", "run.sh"]
```

**Reply by deadcoder0904** [reply]
@viantirreau [this solution is better i think as it was recommended by sharp.](https://github.com/vercel/next.js/discussions/35296#discussioncomment-8626288)

**Comment by deadcoder0904** [comment]
if you are using `pnpm`, you need only 2 steps:

1. put this in your `package.json`:

```json
	"pnpm": {
		"supportedArchitectures": {
			"os": [
				"current"
			],
			"cpu": [
				"x64",
				"arm64"
			]
		}
	}
```

2. install `sharp` normally

```dockerfile
RUN pnpm install --frozen-lockfile --prefer-offline
RUN pnpm install sharp@0.32.6
```

don't forget to remove `sharp` from your `package.json`. you also don't need `ENV_SHARP_PATH` either.

if this is not working, then do `rm -rf node_modules && pnpm install` & it will work.

**Comment by muhammedogz** [comment]
I think easiest solution is installing sharp in docker before running the server.

I completing every step to build my app, after that I am installing sharp before running my server.
Also don't forget to remove sharp from your `package.json`.

<details>
  <summary>Full Docker</summary>
  
  ```Dockerfile
# Install dependencies only when needed
FROM node:20.7.0-alpine AS deps

RUN apk add --no-cache libc6-compat

WORKDIR /app

COPY package.json ./ 

RUN npm i

# Rebuild the source code only when needed
FROM node:20.7.0-alpine AS builder

ENV NEXT_TELEMETRY_DISABLED 1

WORKDIR /app

COPY --from=deps /app/node_modules ./node_modules
COPY . .

RUN npm run build

# Production image, copy all the files and run next
FROM node:20.7.0-alpine AS runner

ENV NEXT_TELEMETRY_DISABLED 1
ENV PORT 3000

WORKDIR /app

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

# You only need to copy next.config.js if you are NOT using the default configuration
COPY --from=builder /app/next.config.js ./
COPY --from=builder /app/public ./public
COPY --from=builder /app/package.json ./package.json
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

# Install sharp
RUN npm i sharp

USER nextjs

EXPOSE 3000

ENTRYPOINT ["node", "server.js"]
  ```
</details>

**Reply by polly3d** [reply]
You save the world. Thanks so much!

**Reply by yanghoxom** [reply]
@muhammedogz sorry, why do we need this?
> Also don't forget to remove sharp from your package.json.

It means we remove it from pkg, and next will not use it anymore, that's why the error does not happen anymore?

**Comment by tihuan** [comment]
For what is worth, we ended up switching from `20-alpine` to `20-bullseye-slim` and it worked

**Comment by ermakovich** [comment]
In case somebody is still struggling with it in 2025, here is the post that might be helpful: [Fix image optimization in NextJS (standalone build)](https://ermakovich.ru/posts/2025-06-30-fix-image-optimization-nextjs-standalone-build/)
